airflow/www/static/js/api/useSetTaskInstanceNotes.ts

```diff
@@ -0,0 +1,114 @@
+  runId: string;
+  taskId: string;
+  mapIndex?: number;
+}
+
+export default function useSetTaskInstanceNotes({
+  dagId, runId, taskId, mapIndex = -1,
+}: Props) {
+  const queryClient = useQueryClient();
+  const errorToast = useErrorToast();
+  // Note: Werkzeug does not like the META URL on dag.html with an integer. It can not put
+  // _MAP_INDEX_ there as it interprets that as the integer. Hence, we pass 0 as the integer.
+  // To avoid we replace other stuff, we add the surrounding strings to the replacement query.
+  const url = (mapIndex >= 0 ? setMappedTaskInstancesNotesURI : setTaskInstancesNotesURI)
+    .replace('_DAG_RUN_ID_', runId)
+    .replace('_TASK_ID_/0/setNote', `_TASK_ID_/${mapIndex}/setNote`)
+    .replace('_TASK_ID_', taskId);
+
+  return useMutation(
+    ['setTaskInstanceNotes', dagId, runId, taskId, mapIndex],
+    (notes: string | null) => axios.patch(url, { notes }),
+    {
+      onSuccess: async (data) => {
+        const notes = (data as API.TaskInstance).notes ?? null;
+
+        const updateMappedInstancesResult = (oldMappedInstances?: API.TaskInstanceCollection) => {
+          if (!oldMappedInstances) {
+            return {
+              taskInstances: [],
+              totalEntries: 0,
+            };
+          }
+          if (mapIndex === undefined || mapIndex < 0) return oldMappedInstances;
+          return {
+            ...oldMappedInstances,
+            taskInstances: oldMappedInstances.taskInstances?.map((ti) => (
+              ti.dagRunId === runId && ti.taskId === taskId && ti.mapIndex === mapIndex
+                ? { ...ti, notes }
+                : ti
+            )),
+          };
+        };
+
+        const updateTaskInstanceResult = (oldTaskInstance?: API.TaskInstance) => {
+          if (!oldTaskInstance) throw new Error('Unknown value...');
+          if (
+            oldTaskInstance.dagRunId === runId
+            && oldTaskInstance.taskId === taskId
+            && (
+              (oldTaskInstance.mapIndex == null && mapIndex < 0)
+              || oldTaskInstance.mapIndex === mapIndex
+            )
+          ) {
+            return {
+              ...oldTaskInstance,
+              notes,
+            };
+          }
+          return oldTaskInstance;
+        };
+        /*
+          This will force a refetch of gridData.
+          Mutating the nested object is quite complicated,
+          we should simplify the gridData API object first
+        */
+        await queryClient.invalidateQueries('gridData');
+
+        if (mapIndex >= 0) {
+          await queryClient.cancelQueries('mappedInstances');
+          queryClient.setQueriesData('mappedInstances', updateMappedInstancesResult);
+        }
+
+        await queryClient.cancelQueries('taskInstance');
+        queryClient.setQueriesData(
+          ['taskInstance', dagId, runId, taskId, mapIndex],
+          updateTaskInstanceResult,
+        );
+      },
+      onError: (error: Error) => errorToast({ error }),
```

Comment:
Shouldn't we rollback the optimistic update in case there is an error ?

Reply:
This isn't an optimistic update. It only runs `onSuccess`. If it was `onMutate`, that would be an optimistic update.

Reply:
You are right thanks